### PR TITLE
EVE networking documentation update

### PR DESCRIPTION
**Pull request by milan-zededa**
Added documentation for recently implemented features (VLANs, LAGs, NI IP routing), but also filling some gaps and providing top-level overview for EVE networking.

### CODE REVIEW

File: docs/CLOCK-SYNCHRONIZATION.md [hunk 1]
```diff
@@ -11,7 +11,7 @@ systems, however, hardware timers are virtualized by the hypervisor, which
 cannot ensure the precision guaranteed by physical hardware devices.
 
 The common and best practice to keep clocks synchronized, for both
-virtualized and non-virtualized systems, is to use the NTP protocol, so
```
Comment (eriknordmark) [inline docs/CLOCK-SYNCHRONIZATION.md:14]: I don't see a link.

File: docs/DEVICE-CONNECTIVITY.md [hunk 2]
```diff
@@ -1,142 +1,483 @@
...
+of the network ports might be both complex and changing over time.
+
+In addition, some network ports might be designated to application usage, or the underlying
+I/O bus controllers (e.g., PCI controllers) be designated for assignment to applications.
+That part of the configuration can also change in run-time.
+
+Any network port configuration changes which might affect the connectivity to the controller
+requires care to avoid permanently losing connectivity to the controller and become unmanageable
+as a result. Thus, it is required to perform testing as part of a configuration change
+and have a way to fall back to a working port configuration. This is accomplished by a logic
```
Comment (eriknordmark) [inline docs/DEVICE-CONNECTIVITY.md:13]: "a logic" -> "logic"

File: docs/DEVICE-CONNECTIVITY.md [hunk 3]
```diff
@@ -1,142 +1,483 @@
...
+network interface representing the device on the data-plane. The latter approach is potentially
+not stable and therefore not recommended. Interface name can change after a reboot if multiple
+devices of the same type are initialized in a different order. This is quite common with cellular
+modems or with ethernet NICs operated by different drivers loaded at random order.
+
+The prerequisite for using a given network IO device inside the EVE domain (i.e. not directly
+assigned to an application), is that the Linux kernel used by EVE is already shipped with
+a suitable (open-source) driver. It may be necessary to adjust the [kernel](https://github.com/lf-edge/eve-kernel)
+options to enable building of the driver.
+For cellular modems this is not required, however. Cellular modems are fairly complex
```
Comment (eriknordmark) [inline docs/DEVICE-CONNECTIVITY.md:50]: Aren't there cases when EVE (and/or eve-kernel) need to be updated to support additional modems?

File: docs/DEVICE-CONNECTIVITY.md [hunk 4]
```diff
@@ -1,142 +1,483 @@
...
+The cost is a number between 0 and 255, with zero (the default) implying free,
+and less preferred ports being assigned a higher cost. Multiple ports can be assigned
+the same cost, in which case once failover has happened to cost N, then all uplink ports
+with cost N will be used for load spreading of the management traffic.
+
+Furthermore, one can set [network.download.max.cost](CONFIG-PROPERTIES.md) to a number N
+if it is acceptable for EVE to perform (potentially large) downloads of content and images
+when having failed over to uplink ports with cost N (or less).
+
+The cost is new and will replace the (now obsolete) `free`/`freeUplink` way to specify
```
Comment (eriknordmark) [inline docs/DEVICE-CONNECTIVITY.md:157]: "new" and "will" might have been true in the past. 

Either we should completely remove the free/freeUplink from the documentation (which definitely makes sense if those old releases are no longer in use), or at least change the tense to say something like (Prior to the introduction of the cost integer there were free/freeUplink booleans which EVE still looks at to provide compatibility for old controllers.)

**Comment by rene** [comment]
@milan-zededa , very nice work! I'm going through the docs yet, will come back to you if I have any questions...

File: docs/DEVICE-CONNECTIVITY.md [hunk 5]
```diff
@@ -1,142 +1,483 @@
...
+
+A network proxy may operate *transparently*, meaning that clients (EVE devices, user apps)
+making HTTP/HTTPS/… requests are not aware of its presence. When a client makes an HTTP(S)
+request, it addresses it directly to the destination (e.g. EVE controller), but a transparent
+proxy will capture it (without the client knowing about it) and apply a proxying action.
+The action may depend on the destination URL. For example, the proxy may be configured to drop
+requests heading to some specific domains, but allow others.
+
+Additionally, the proxy may break the TLS tunnel into two (client-proxy, proxy-destination)
+and act as a MITM, reading the otherwise encrypted traffic. This is done by proxy generating
```
Comment (eriknordmark) [inline docs/DEVICE-CONNECTIVITY.md:183]: I don't know if we already spell out MiTM - if not it makes sense to do e.g. "Man-in-the-Middle (MITM)" the first time we use it.

File: docs/DEVICE-CONNECTIVITY.md [hunk 6]
```diff
@@ -1,142 +1,483 @@
...
+Information about TLS tunnels opened and peer certificates presented is also part
+of the [network tracing](#netdump-and-nettrace).
 
-To handle different types of connectivity towards the controller EVE supports both load spreading and failover when it comes to maintaining connectivity to the controller.
+### Explicit network proxy
 
-## Load spreading and failover
+Presence of *explicit* proxy is known to the client making HTTP/HTTPS/… requests.
+The client will ask the proxy explicitly to proxy a given request - e.g. make connection
+for me to the controller and perform a GET request for path xyz, then return the result
```
Comment (eriknordmark) [inline docs/DEVICE-CONNECTIVITY.md:209]: I don't know how much detail you want to get into here, but it could be a POST or a GET. An "URL" might be a better term than "path" here.

But less detail would be e.g., "will explicitly ask the proxy to make a HTTPS request to a specified URL."

File: docs/DEVICE-CONNECTIVITY.md [hunk 7]
```diff
@@ -1,142 +1,483 @@
...
+back to me. In this case, the client is configured with `host:port` proxy address for every
```
Comment (eriknordmark) [inline docs/DEVICE-CONNECTIVITY.md:210]: Explicitly host:port is only one way since we also support an explicltly configured PAC file, and dynamic discovery and fetch of a PAC file.

File: docs/DEVICE-CONNECTIVITY.md [hunk 8]
```diff
@@ -1,142 +1,483 @@
...
 ### Fresh install
 
-When EVE is installed using a generic EVE installer without device bootstrap configuration included (see [CONFIG.md](./CONFIG.md), section "Bootstrap configuration"), it only has the last-resort configuration available (see "Last resort" section below). This is sufficient to connect to the controller if
+When EVE is installed using a generic EVE installer without device bootstrap configuration
+included (see [CONFIG.md](./CONFIG.md), section "Bootstrap configuration"), it only has
+the last-resort configuration available (see "Last resort" section below). This is sufficient
+to connect to the controller if
 
-- DHCP is enabled on one of the Ethernet ports
+- DHCP is enabled on one of the uplink Ethernet ports
```
Comment (eriknordmark) [inline docs/DEVICE-CONNECTIVITY.md:283]: Nit, but why did you add "uplink" here? If there is no bootstrap configuration then all Ethernet ports are tried.

File: docs/NETWORKING.md [hunk 9]
```diff
@@ -0,0 +1,156 @@
+# EVE Networking (Overview)
```
Comment (eriknordmark) [inline docs/NETWORKING.md:1]: Thank you for this overview document.

With it plus APP-CONNECTIVITY.md it seems like we've subsumed (with more details) what's in NETWORK-MODELS.md. Can you check whether that is the case and if so remove it?

File: docs/APP-CONNECTIVITY.md [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+
+Despite Switch NI operating only at the L2 layer, the L3/L4 [network ACLs](NETWORK-ACLS.md)
+are still supported and must be configured to allow anything beyond just DNS and DHCP to pass.
+Since there is no NAT between applications and external endpoints, properly configured
+inbound ACL rules are that much more important.
+
+Metadata HTTP server is run for switch network instance only if it has uplink port with
+an IP address.
+
+### Uplink Port
```
Comment (eriknordmark) [inline docs/APP-CONNECTIVITY.md:210]: I know we call it "uplink" in the API, but this is really the EVE management port(s).
I wonder if that terminology would be more clear throughout since e.g., eth3 could be pointing up towards the Internet but not be used for EVE management traffic. Thoughts?

**Review by eriknordmark — COMMENTED** [review]
Very good additions - I've added a few minor comments separately.

@naiming-zededa can you review the routing parts in particular - there are some images etc.

**Comment by naiming-zededa** [comment]
> Very good additions - I've added a few minor comments separately.
> 
> @naiming-zededa can you review the routing parts in particular - there are some images etc.

will do. thanks for this detailed documentation.

File: docs/APP-CONNECTIVITY.md [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+or "drop" mark. Based on the mark and the src/dst addresses, IP rules either send the packet
+into the dummy blackole interface (dropped), or route the packet according to the NI routing
+table. If the destination is external, packet will be transmitted out through the uplink port.
+If the destination is another app (inside the same NI), packet's dst MAC address will be set
+to the dst app MAC, and it will be forwarded again through the same bridge into the corresponding
+VIF. Finally, the packet is transmitted from the host network stack into the guest network
+stack of the destination application, where it will be processed for local delivery.
+The diagram below depicts all these packet-flow stages:
+
+![packet-flow](./images/eve-app-packet-flow.png)
```
Comment (naiming-zededa) [inline docs/APP-CONNECTIVITY.md:90]: Will be good to add anther diagram for app packets inbound (the returned traffic, and also port-mapped packets)

File: docs/APP-CONNECTIVITY.md [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+feature-sets and performance considerations.
+
+## Application network interfaces
+
+Network connectivity is enabled for an application by configuring one or more network adapters.
+These consists of [virtual network interfaces](#virtual-network-interfaces), connecting
+application with [network instances](#network-instances), and directly assigned
+[physical NIC](#physical-network-ports) or [SR-IOV virtual functions](#sr-iov-vfs).
+The order of network adapter configurations inside `AppInstanceConfig` matters, because this
+is the same order at which these network adapters will be presented to the application by VMM.
```
Comment (naiming-zededa) [inline docs/APP-CONNECTIVITY.md:14]: Probably should expand the 'VMM', in EVE doc, it can also be 'VM Monitor'.

File: docs/APP-CONNECTIVITY.md [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+Most of the edge deployments will deploy applications that will have connectivity to both
+WAN (Internet) and LAN (e.g. shop floor, machine floor).
+There may be a single WAN interface and multiple LAN interfaces. With the IP route
+configurability of (Local) Network instances provided by EVE, we can automatically (zero-touch)
+provision routing for applications that have access to both the Internet and one or more LANs.
+This is possible because EVE allows to:
+
+* use DHCP (option 121) to propagate the *connected routes* to applications (routes for external
+  networks that NI uplinks are connected to)
+* configure a set of *static IP routes* for a network instance and have them propagated
```
Comment (naiming-zededa) [inline docs/APP-CONNECTIVITY.md:261]: May add 'also by DHCP'

**Review by naiming-zededa — APPROVED** [review]
LGTM. Very good documents. A couple of comments. thanks.

File: docs/APP-CONNECTIVITY.md [hunk 14]
```diff
@@ -0,0 +1,299 @@
...
+the connected routes, which, when enabled, EVE will automatically propagate into the application.
+
+For external networks one or more routing hops away (i.e. not directly connected to the device),
+user is able to configure static IP routes and let EVE propagate them to the application
+also using DHCP. Static routes are part of the network instance configuration. User configures
+destination IP subnets which the network instance will become the gateway for (from the app
+perspective). In the example below, the propagated static route is highlighted with the green
+color.
+
+![ni-ip-routing](./images/eve-ip-routing.png)
```
Comment (naiming-zededa) [inline docs/APP-CONNECTIVITY.md:275]: Wondering if this 'eve-ip-routing' should add one 'default route' in the diagram to one of the external port

**Review by uncleDecart — COMMENTED** [review]
Thank you @milan-zededa for doing this. It's really good. I'll need to re-read it couple of times, so more comments will follow

File: docs/APP-CONNECTIVITY.md [hunk 15]
```diff
@@ -0,0 +1,299 @@
...
+
+SR-IOV NIC VFs (virtual functions) combine the resource-sharing advantages of virtual interfaces
+with performance benefits of direct assignments.
+In this case, the virtualization of the physical resource (NIC) is done by the hardware itself.
+EVE OS manages the PF (physical function) inside the host, creates virtual functions based
+on the configuration (device model), and performs direct assignment of these VFs into applications
+(using the very same technologies as for physical ports).
+
+The main difference between directly assigned VF and a physical port from the application
+perspective, is that virtual function requires a different "VF" driver variant to operate
```
[the remark this answers is not in the document]
Comment (milan-zededa) [reply docs/APP-CONNECTIVITY.md:44]: "for Guest" - do you mean container apps? Is the driver part of the shim VM that we create for container?
I will mention it if it is the case.
VM apps and their content is fully under the user control. Driver for Intel NIC SR-IOV VF may not be included.

File: docs/DEVICE-CONNECTIVITY.md [hunk 16]
```diff
@@ -1,142 +1,483 @@
...
+as a result. Thus, it is required to perform testing as part of a configuration change
+and have a way to fall back to a working port configuration. This is accomplished by a logic
+to test connectivity to the controller, and persistently maintaining a list of network
+configurations, including the latest and possibly one or more fallback configurations.
+
+## Network adapters
+
+Network adapter denotes network configuration (MTU, MAC, IP, VLAN, etc.) and logical attributes
+(logical & interface names, cost, usage) attached to a physical network port.
+As show in the diagram below, there is a hierarchy of different types of network adapters
```
Comment (uncleDecart) [inline docs/DEVICE-CONNECTIVITY.md:21]: As **shown** in diagram below?

File: docs/DEVICE-CONNECTIVITY.md [hunk 17]
```diff
@@ -1,142 +1,483 @@
...
+
+Network adapters use logical labels for identification and for reference within the hierarchical
+layers. Unlike higher-level objects like apps or network instances that use UUIDs, these labels
+are user-defined, simpler human-readable strings, which are only required to be unique within
+the edge-node, not globally.
+
+### Physical network ports
+
+EVE supports ethernet NICs, WiFi modules and cellular modems.
+Support for [CAN devices is a work in-progress](https://zededa.atlassian.net/wiki/spaces/EVE/pages/2163539969/Introducing+CAN+support+on+EVE).
```
Comment (uncleDecart) [inline docs/DEVICE-CONNECTIVITY.md:35]: This should be link to LF-Edge wiki, or?

File: docs/NETWORKING.md [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+
+Diagram below depicts all the components of the data-plane and how they all fit together.
+Also shown is the separation of the management plane between EVE microservices.
+
+![networking-impl](./images/eve-networking-impl.png)
+
+To dive deeper into the EVE networking, including the implementation details, see the next section
+for further directions.
+
+## Where to go next
```
Comment (uncleDecart) [inline docs/NETWORKING.md:146]: Should we add ECO-METADATA link here?

File: docs/APP-CONNECTIVITY.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (uncleDecart) [reply docs/APP-CONNECTIVITY.md:44]: I meant deployed VM, for VM created VF would be "special" PCI device, meaning VF driver will know how to interpret it and how to set it up.

File: docs/APP-CONNECTIVITY.md [hunk 19]
```diff
@@ -0,0 +1,299 @@
...
+Network instance (NI for short) is a virtual switch deployed inside an edge-device,
+allowing applications (virtual machines or containers) to communicate with one another over
+a virtual network. Network instance represents a single network segment. Depending on the type
+and configuration, it is either internal to the edge-node or an extension of an external network
+segment.
+
+Network instances are a feature of the EVE OS host networking build using various software tools.
+Applications connect to network instances using [virtual network interfaces](#virtual-network-interfaces).
+Directly assigned physical network ports or SR-IOV VFs bypass the host networking, thus cannot
+be part of network instances. Apart from VIFs, network instance can be connected to a physical
```
[the remark this answers is not in the document]
Comment (milan-zededa) [reply docs/APP-CONNECTIVITY.md:103]: I only document what we support today.

File: docs/DEVICE-CONNECTIVITY.md [hunk 20]
```diff
@@ -1,142 +1,483 @@
...
+to be used by multiple virtual machines simultaneously without the mediation from the hypervisor.
+The actual physical device is known as a Physical Function (PF) while the others are known
+as Virtual Functions (VF). With this feature, the hypervisor can attach one or more VFs
+to a Virtual Machine so that the guest can then use the device as if it was directly
+assigned.
+
+To use this feature with EVE, first you need to check that (PF) driver for your SR-IOV-enabled
+NIC is already shipped with EVE OS. With proprietary drivers this is going to be a problem.
+Until we add a feature to deploy proprietary drivers for the host inside a VM just like
+applications (feature proposal knows as "driver domain"), EVE users are limited to open-source
```
[the remark this answers is not in the document]
Comment (milan-zededa) [reply docs/DEVICE-CONNECTIVITY.md:68]: Which driver/NIC was tested?

[the remark this answers is not in the document]
Comment (uncleDecart) [reply docs/DEVICE-CONNECTIVITY.md:68]: Intel i350

File: docs/CLOCK-SYNCHRONIZATION.md (hunk 1 above)
Comment (milan-zededa) [reply docs/CLOCK-SYNCHRONIZATION.md:14]: Fixed

File: docs/DEVICE-CONNECTIVITY.md (hunk 2 above)
Comment (milan-zededa) [reply docs/DEVICE-CONNECTIVITY.md]: Fixed

File: docs/DEVICE-CONNECTIVITY.md (hunk 3 above)
Comment (milan-zededa) [reply docs/DEVICE-CONNECTIVITY.md:50]: I added clarification that we support only modems connected over USB (and not over PCI). In that case, the modem interface is unified and we control any modem with the same set of drivers.

File: docs/DEVICE-CONNECTIVITY.md (hunk 5 above)
Comment (milan-zededa) [reply docs/DEVICE-CONNECTIVITY.md]: Done

File: docs/DEVICE-CONNECTIVITY.md (hunk 6 above)
Comment (milan-zededa) [reply docs/DEVICE-CONNECTIVITY.md]: Updated the sentence as suggested.

File: docs/DEVICE-CONNECTIVITY.md (hunk 7 above)
Comment (milan-zededa) [reply docs/DEVICE-CONNECTIVITY.md]: I have restructured this text. Now the proxy address auto-discovery is under explicit proxy.

File: docs/DEVICE-CONNECTIVITY.md (hunk 8 above)
Comment (milan-zededa) [reply docs/DEVICE-CONNECTIVITY.md:287]: These are the requirements for working connectivity with the cloud when lastresort is used. So it is not about what EVE will try but what must be provided.

File: docs/NETWORKING.md (hunk 9 above)
Comment (milan-zededa) [reply docs/NETWORKING.md:1]: True, I checked and there is now no added value in NETWORK-MODELS.md
I have removed the file and links to it.

File: docs/APP-CONNECTIVITY.md (hunk 10 above)
Comment (milan-zededa) [reply docs/APP-CONNECTIVITY.md:235]: I agree that "management port" would be better. However, "uplink" is quite infiltrated now across the code base, in the API and in our (open and proprietary) documentation. I guess we are stuck with it now.

File: docs/APP-CONNECTIVITY.md (hunk 11 above)
Comment (milan-zededa) [reply docs/APP-CONNECTIVITY.md:95]: Added

File: docs/APP-CONNECTIVITY.md (hunk 12 above)
Comment (milan-zededa) [reply docs/APP-CONNECTIVITY.md]: Done

File: docs/APP-CONNECTIVITY.md (hunk 13 above)
Comment (milan-zededa) [reply docs/APP-CONNECTIVITY.md:286]: Done

File: docs/APP-CONNECTIVITY.md (hunk 14 above)
Comment (milan-zededa) [reply docs/APP-CONNECTIVITY.md:300]: Added

File: docs/APP-CONNECTIVITY.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (milan-zededa) [reply docs/APP-CONNECTIVITY.md:45]: Added clarification that Intel NIC I350 with igb(vf) driver is the only one tested and confirmed to work with EVE.

File: docs/DEVICE-CONNECTIVITY.md (hunk 16 above)
Comment (milan-zededa) [reply docs/DEVICE-CONNECTIVITY.md]: Fixed

File: docs/DEVICE-CONNECTIVITY.md (hunk 17 above)
Comment (milan-zededa) [reply docs/DEVICE-CONNECTIVITY.md]: Yes, fixed.

File: docs/NETWORKING.md (hunk 18 above)
Comment (milan-zededa) [reply docs/NETWORKING.md:145]: Added

File: docs/DEVICE-CONNECTIVITY.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (milan-zededa) [reply docs/DEVICE-CONNECTIVITY.md]: Done

File: docs/DEVICE-CONNECTIVITY.md (hunk 3 above)
Comment (eriknordmark) [reply docs/DEVICE-CONNECTIVITY.md:50]: Ok

File: docs/DEVICE-CONNECTIVITY.md (hunk 8 above)
Comment (eriknordmark) [reply docs/DEVICE-CONNECTIVITY.md:287]: Oh - I misread the sentence above it.

**Review by eriknordmark — APPROVED** [review]
LGTM
